budgetportal/models.py

```diff
@@ -168,6 +171,63 @@ def get_resources(self):
                     })
         return resources
 
+    def get_govt_functions(self):
+        return GovtFunction.objects.filter(programme__department=self).distinct()
+
+    def get_financial_year(self):
+        return self.government.sphere.financial_year
+
+    def _get_financial_year_query(self):
+        return '+vocab_financial_years:"%s"' % self.government.sphere.financial_year.slug
+
+    def _get_government_query(self):
+        if self.government.sphere.slug == 'provincial':
+            return '+vocab_provinces:"%s"' % self.government.name
+        else:
+            return none_selected_query('vocab_provinces')
+
+    def _get_functions_query(self):
+        function_names = [f.name for f in self.get_govt_functions()]
+        ckan_tag_names = [re.sub('[^\w -]', '', n) for n in function_names]
+        if len(ckan_tag_names) == 0:
+            return none_selected_query('vocab_functions')
+        else:
+            options = ['+vocab_functions:"%s"' % n for n in ckan_tag_names]
+            return '+(%s)' % ' OR '.join(options)
+
+    def get_contributed_datasets(self):
+        # We use an OrderedDict like an Ordered Set to ensure we include each
+        # match just once, and at the highest rank it came up in.
+        datasets = OrderedDict()
+
+        fq_org = '-organization:"national-treasury"'
+        fq_year = self._get_financial_year_query()
+        fq_sphere = '+vocab_spheres:"%s"' % self.government.sphere.slug
+        fq_government = self._get_government_query()
+        fq_functions = self._get_functions_query()
+        fq_no_functions = none_selected_query('vocab_functions')
+        queries = [
```

Comment:
we might need to adjust these in future or add more permutations but this is a decent start